[Security][SecurityBundle] Add encryption support to OIDC tokens

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #50441
| License       | MIT

The changes add encryption support to OpenID Connect (OIDC) tokens in the Symfony Security Bundle. This is useful in making the application more secure. They also ensure the tokens are correctly decrypted and validated before use. Additionally, tests have been expanded to cover these new scenarios.

```yaml
security:
    firewalls:
        main:
            pattern: ^/
            access_token:
                token_handler:
                    oidc:
                        ...
                        encryption:
                            enabled: true
                            algorithms: [...]
                            keyset: '{"keys": [{...}]}'
```

Spomky commented:
Ping @vincentchalamon

vincentchalamon commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/AccessToken/config_oidc.yml:
Shouldn't you create a `config_oidc_encryption.yml` file instead, to keep this one working out of encryption?

vincentchalamon commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/AccessToken/config_oidc.yml:
Shouldn't you create a `config_oidc_encryption.yml` file instead, to keep this one working out of encryption?

Spomky replied:
With the configuration, encryption is not mandatory.
Because there is an option to require only JWE. I will create a second config file.

chalasr replied:
A separate test case for the no encryption case would be better indeed

Spomky replied:
I added a specific test where a valid JWS is sent and a 401 status code is returned because a JWE is required

vincentchalamon commented on src/Symfony/Component/Security/Http/AccessToken/Oidc/OidcTokenHandler.php:
```diff
@@ -37,10 +41,13 @@
     public function __construct(
         private Algorithm|AlgorithmManager $signatureAlgorithm,
-        private JWK|JWKSet $jwkset,
+        private JWK|JWKSet $signatureKeyset,
```
I'm wondering if changing the argument/property name may cause a BC break, particularly about named parameters.

Spomky replied:
AFAIK, there is no BC promise regarding the argument names (except for attributes).
@nicolas-grekas: do you confirm? Is it possible to change the argument names here?

Spomky replied:
See[ `[10]` of the BC promise](https://symfony.com/doc/current/contributing/code/bc.html#note-10)

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:
```diff
@@ -41,6 +41,22 @@ public function create(ContainerBuilder $container, string $id, array|string $co
             ->replaceArgument(0, $config['keyset'])
         );
+
+        if ($config['encryption']['enabled'] === true) {
```
is the boolean comparison required?
comparing a bool to a bool to get a bool is to much :)

Spomky replied:
Fixed👍🏼

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:
```diff
@@ -41,6 +41,22 @@ public function create(ContainerBuilder $container, string $id, array|string $co
+            $algorithmManager = (new ChildDefinition('security.access_token_handler.oidc.encryption'))
+            ->replaceArgument(0, $config['encryption']['algorithms']);
+            $keyset = (new ChildDefinition('security.access_token_handler.oidc.jwkset'))
+            ->replaceArgument(0, $config['encryption']['keyset']);
```
`->` lines should be indented

Spomky replied:
Indentation added

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator_access_token.php:
```diff
@@ -135,5 +144,49 @@
+            ])
+
+        ->set('security.access_token_handler.oidc.encryption.RSAOAEP', RSAOAEP::class)
+        ->tag('security.access_token_handler.oidc.encryption_algorithm')
```
all `->tag` should be indented

Spomky replied:
Indentation added

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/AccessTokenTest.php:
```diff
@@ -408,4 +403,90 @@ public function testCasSuccess()
+        ];
+    }
+
+    public function invalidAccessTokens(): array
```
data providers should be static

Spomky replied:
Done

nicolas-grekas commented on src/Symfony/Component/Security/Http/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Pass the current token to the `checkPostAuth()` method of user checkers
  * Deprecate argument `$secret` of `RememberMeAuthenticator`
+* Add encryption support to `OidcTokenHandler` (JWE)
```
```suggestion
 * Add encryption support to `OidcTokenHandler` (JWE)
```

Spomky replied:
Fixed

nicolas-grekas reviewed (commented):
Is there no XSD to update?

OskarStark commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator_access_token.php:
```diff
@@ -135,5 +144,49 @@
 
         ->set('security.access_token_handler.oidc.signature.PS512', PS512::class)
             ->tag('security.access_token_handler.oidc.signature_algorithm')
+
+
```
```suggestion
```

Spomky replied:
Thanks. Done

Spomky commented:
> Is there no XSD to update?

@nicolas-grekas I'm not sure what to change.

chalasr commented:
@Spomky The new options need to be added to SecurityBundle/Resources/config/schema/security-1.0.xsd

Spomky commented:
> @Spomky The new options need to be added to SecurityBundle/Resources/config/schema/security-1.0.xsd

Hello @chalasr,
I modified the schema. Note that I am not sure it is fine of not (I was not aware of it before your comment).
Regards.

chalasr commented on src/Symfony/Component/Security/Http/AccessToken/Oidc/OidcTokenHandler.php:
```diff
@@ -111,4 +97,92 @@ public function getUserBadgeFrom(string $accessToken): UserBadge
+            }
+
+            return $payload;
+        } catch (\Throwable $e) {
```
Could it catch more precise exception types? Catch-all should generally be avoided

Spomky commented:
Hi,

I am not sure how to fix the failing tests.
The annotation `@requires extension openssl` is present, but it looks like the dataProvider is called.
Should I refactor these tests?

chalasr commented:
You may need to bump the security-http dependency to ^7.3 in security-bundle

Spomky committed:
[Security] OAuth2 Introspection Endpoint (RFC7662)

In addition to the excellent work of @vincentchalamon #48272, this PR allows getting the data from the OAuth2 Introspection Endpoint. This endpoint is defined in the [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662). It returns the following information that is used to retrieve the user:

* If the access token is active
* A set of claims that are similar to the OIDC one, including the `sub` or the `username`.

Spomky commented:
Many thaks.
Took me a while to understand the problem with the remaining failing tests.😅
Sorry for the noise. All green now.

fabpot commented:
Thank you @Spomky.
